File: android/app/src/main/java/com/openland/spacex/store/SpaceXStore.kt
package com.openland.spacex.store

import android.content.Context
import com.openland.spacex.OperationDefinition
import com.openland.spacex.OperationKind
import com.openland.spacex.persistence.SpaceXPersistence
import com.openland.spacex.utils.DispatchQueue
import com.openland.spacex.utils.trace
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicInteger

sealed class QueryReadResult {
    class Value(val value: JSONObject) : QueryReadResult()
    object Missing : QueryReadResult()
}

sealed class QueryReadAndWatchResult {
    class Value(val value: JSONObject) : QueryReadAndWatchResult()
    object Missing : QueryReadAndWatchResult()
    object Updated : QueryReadAndWatchResult()
}

private class Subscription(val id: Set<String>, val callback: () -> Unit)
private class ReadRequest(val missing: MutableSet<String>, val callback: () -> Unit)


class SpaceXStore(context: Context, name: String?) {
    companion object {
        const val ROOT_QUERY = "ROOT_QUERY"
    }

    private val normalizationQueue = DispatchQueue("normalization")
    private val storeQueue = DispatchQueue("store")

    // Store
    private val store = RecordStore()

    // Persistence
    private val pendingReadRequests = mutableListOf<ReadRequest>()
    private val requested = mutableSetOf<String>()
    private val persistence = SpaceXPersistence(context, name)
    private var isWriting = false
    private val pendingWrite = mutableMapOf<String, Record>()

    // Bus
    private var nextSubscriptionId = AtomicInteger(1)
    private var subscriptions = mutableMapOf<Int, Subscription>()

    fun mergeResponse(operation: OperationDefinition, variables: JSONObject, data: JSONObject, queue: DispatchQueue, callback: () -> Unit) {
        this.normalizationQueue.async {
            val rootCacheKey = if (operation.kind == OperationKind.QUERY) ROOT_QUERY else null
            val normalized = normalizeResponse(rootCacheKey, operation.selector, variables, data)
            this.merge(normalized, queue, callback)
        }
    }

    fun merge(recordSet: RecordSet, queue: DispatchQueue, callback: () -> Unit) {
        this.storeQueue.async {
            prepareMerge(recordSet) {

                // Merge data in RAM store
                val changed = store.merge(recordSet)

                // Write data to disk
                if (changed.isNotEmpty()) {
                    val toWrite = RecordSet(changed.mapValues { store.read(it.key) })
                    this.persistenceWrite(toWrite)
                }

                // Notify watchers
                if (changed.isNotEmpty()) {
                    val triggered = mutableListOf<Subscription>()
                    val keys = mutableSetOf<String>()
                    for (r in changed) {
                        for (f in r.value.fields) {
                            keys.add(r.key + "." + f)
                        }
                    }
                    for (s in subscriptions) {
                        if (s.value.id.any { keys.contains(it) }) {
                            triggered.add(s.value)
                        }
                    }
                    for (t in triggered) {
                        t.callback()
                    }
                }

                // Invoke callback
                queue.async {
                    callback()
                }
            }
        }
    }

    fun readQuery(operation: OperationDefinition, variables: JSONObject, queue: DispatchQueue, callback: (result: QueryReadResult) -> Unit) {
        this.storeQueue.async {
            prepareRead(operation, variables) {
                // Read from in-memory store
                val res = readRootFromStore(SpaceXStore.ROOT_QUERY, store, operation.selector, variables)

                // Notify about result
                if (res.first) {
                    queue.async {
                        callback(QueryReadResult.Value(res.second!!))
                    }
                } else {
                    queue.async {
                        callback(QueryReadResult.Missing)
                    }
                }
            }
        }
    }

    fun readQueryAndWatch(operation: OperationDefinition, variables: JSONObject, queue: DispatchQueue, callback: (result: QueryReadAndWatchResult) -> Unit) {
        this.storeQueue.async {
            prepareRead(operation, variables) {
                // Read from in-memory store
                val res = readRootFromStore(SpaceXStore.ROOT_QUERY, store, operation.selector, variables)

                // If missing
                if (!res.first) {
                    queue.async {
                        callback(QueryReadAndWatchResult.Missing)
                    }
                    return@prepareRead
                }
                // If exists
                queue.async {
                    callback(QueryReadAndWatchResult.Value(res.second!!))
                }

                // Calculate keys
                // TODO: Optimize
                val normalized = normalizeResponse(SpaceXStore.ROOT_QUERY, operation.selector, variables, res.second!!)
                val keys = mutableSetOf<String>()
                for (r in normalized.records) {
                    for (f in r.value.fields) {
                        keys.add(r.key + "." + f.key)
                    }
                }

                // Subscribe for first notification
                val subscId = nextSubscriptionId.getAndIncrement()
                this.subscriptions[subscId] = Subscription(keys) {
                    this.subscriptions.remove(subscId)
                    queue.async {
                        callback(QueryReadAndWatchResult.Updated)
                    }
                }
            }
        }
    }

    fun close() {
        // TODO: Implement
    }

    private fun prepareMerge(recordSet: RecordSet, callback: () -> Unit) {
        val missing = trace("prepareMerge") { recordSet.records.keys.filter { !store.isInMemory(it) } }
        if (missing.isNotEmpty()) {
            persistenceRead(missing.toSet()) {
                callback()
            }
        } else {
            callback()
        }
    }

    private fun prepareRead(operation: OperationDefinition, variables: JSONObject, callback: () -> Unit) {
        val missing = trace("collectMissingKeys") { collectMissingKeysRoot(SpaceXStore.ROOT_QUERY, store, operation.selector, variables) }
        if (missing.isNotEmpty()) {
            persistenceRead(missing) {
                prepareRead(operation, variables, callback)
            }
        } else {
            callback()
        }
    }


    private fun persistenceRead(keys: Set<String>, callback: () -> Unit) {
        pendingReadRequests.add(ReadRequest(keys.toMutableSet(), callback))
        val filtered = keys.filter { !requested.contains(it) }
        if (filtered.isNotEmpty()) {
            doLoad(filtered.toSet())
        }
    }

    private fun persistenceWrite(records: RecordSet) {
        this.storeQueue.async {
            pendingWrite.putAll(records.records)
            doWriteIfNeeded()
        }
    }

    private fun doWriteIfNeeded() {
        if (pendingWrite.isNotEmpty() && !isWriting) {
            isWriting = true
            val toWrite = RecordSet(pendingWrite.toMap())
            pendingWrite.clear()
            persistence.saveRecords(toWrite, this.storeQueue) {
                isWriting = false
                doWriteIfNeeded()
            }
        }
    }

    private fun doLoad(keys: Set<String>) {
        requested.addAll(keys)
        persistence.loadRecords(keys, this.storeQueue) {

            // Some health check
            for (k in keys) {
                if (!it.records.containsKey(k)) {
                    throw Error("Key $k was not loaded from persistence!")
                }
            }

            store.loaded(it)

            // Notify
            for (p in pendingReadRequests) {
                p.missing.removeAll(keys)
            }
            val ready = pendingReadRequests.filter { it.missing.isEmpty() }
            pendingReadRequests.removeAll { it.missing.isEmpty() }
            for (r in ready) {
                r.callback()
            }
        }

    }
}

File: packages/openland-graphql/spacex/store/SpaceXStore.ts
import { SpaceXPersistence } from './../persistence/SpaceXPersistence';
import { OperationDefinition } from './../types';
import { RecordStore, RecordSet, Record } from './RecordStore';
import { normalizeResponse } from './normalize';
import { readRootFromStore } from './readFromStore';
import { collectMissingKeysRoot } from './collectMissingKeys';

const ROOT_QUERY = 'ROOT_QUERY';

export type QueryAndWatchArg = { type: 'updated' } | { type: 'missing' } | { type: 'value', value: any };

export class SpaceXStore {
    private readonly store = new RecordStore();
    private readonly persistence = new SpaceXPersistence();

    // Persistence Read
    private readonly requested = new Set<string>();
    private readonly pendingReadRequests: { missing: Set<string>, callback: () => void }[] = [];

    // Persistence Write
    private isWriting = false;
    private readonly pendingWriteRequests: Map<string, Record> = new Map();

    // Subscriptions
    private nextSubscriptionId = 1;
    private subscriptions = new Map<number, { keys: Set<string>, callback: () => void }>();

    //
    // Merge to the store
    //

    mergeResponse = async (operation: OperationDefinition, variables: any, data: any) => {
        let rootCacheKey = operation.kind === 'query' ? ROOT_QUERY : null;
        let normalized = normalizeResponse(rootCacheKey, operation.selector, variables, data);
        await this.merge(normalized);
    }

    private merge = async (recordSet: RecordSet) => {
        await this.prepareMerge(recordSet);

        let changed = this.store.merge(recordSet);
        let isEmpty = Object.keys(changed).length === 0;

        // Persist
        if (!isEmpty) {
            let toWrite: RecordSet = {};
            for (let ch of Object.keys(changed)) {
                toWrite[ch] = this.store.read(ch);
            }
            this.persistenceWrite(toWrite);
        }

        // Notify
        if (!isEmpty) {
            let triggered: { keys: Set<string>, callback: () => void }[] = [];
            let keys = new Set<string>();
            for (let ch of Object.keys(changed)) {
                for (let f of Object.keys(changed[ch])) {
                    keys.add(ch + '.' + f);
                }
            }
            for (let s of this.subscriptions) {
                let isTriggered = false;
                for (let k of s[1].keys) {
                    if (keys.has(k)) {
                        isTriggered = true;
                        break;
                    }
                }
                if (isTriggered) {
                    triggered.push(s[1]);
                }
            }
            for (let f of triggered) {
                f.callback();
            }
        }
    }

    readQuery = async (operation: OperationDefinition, variables: any) => {
        await this.prepareRead(operation, variables);

        let root = readRootFromStore(ROOT_QUERY, this.store, operation.selector, variables);

        if (root.result) {
            return { result: true, value: root.value! };
        } else {
            return { result: false };
        }
    }

    readQueryAndWatch = (operation: OperationDefinition, variables: any, callback: (c: QueryAndWatchArg) => void) => {
        (async () => {
            await this.prepareRead(operation, variables);

            let root = readRootFromStore(ROOT_QUERY, this.store, operation.selector, variables);

            if (!root.result) {
                callback({ type: 'missing' });
                return;
            }

            // Calculate keys
            // TODO: Optimize
            let normalized = normalizeResponse(ROOT_QUERY, operation.selector, variables, root.value!);
            let keys = new Set<string>();
            for (let r of Object.keys(normalized)) {
                for (let f of Object.keys(normalized[r])) {
                    keys.add(r + '.' + f);
                }
            }

            // Subscribe
            let subsId = this.nextSubscriptionId++;
            this.subscriptions.set(subsId, {
                keys, callback: () => {
                    this.subscriptions.delete(subsId);
                    callback({ type: 'updated' });
                }
            });

            // Current value
            callback({ type: 'value', value: root.value! });
        })();
    }

    //
    // Read from Persistence
    //

    private readonly prepareMerge = async (recordSet: RecordSet) => {
        let missingIds: Set<string> = new Set(); // Should we avoid allocation?
        for (let k of Object.keys(recordSet)) {
            if (!this.store.isInMemory(k)) {
                missingIds.add(k);
            }
        }
        if (missingIds.size > 0) {
            await this.persistenceRead(missingIds);
        }
    }

    private readonly prepareRead = async (operation: OperationDefinition, variables: any) => {
        let missing = collectMissingKeysRoot(ROOT_QUERY, this.store, operation.selector, variables);
        if (missing.size > 0) {
            await this.persistenceRead(missing);
            await this.prepareRead(operation, variables);
        }
    }

    private readonly persistenceRead = async (keys: Set<string>) => {
        let r: () => void;
        let p = new Promise<void>((resolver) => r = resolver);
        this.pendingReadRequests.push({ missing: keys, callback: r! });

        let filtered: Set<string> = new Set();
        for (let k of keys) {
            if (!this.requested.has(k)) {
                filtered.add(k);
            }
        }

        if (filtered.size > 0) {
            this.doLoad(filtered);
        }

        await p;
    }

    private readonly doLoad = async (keys: Set<string>) => {
        for (let k of keys) {
            this.requested.add(k);
        }

        let loaded = await this.persistence.loadRecords(keys);

        // Some health check
        for (let k of keys) {
            if (!loaded[k]) {
                throw Error('Key ' + k + ' was not loaded from persistence!');
            }
        }

        this.store.loaded(loaded);

        // Notify
        for (let p of this.pendingReadRequests) {
            for (let k of keys) {
                p.missing.delete(k);
            }
        }

        let ready = this.pendingReadRequests.filter((v) => v.missing.size === 0);
        for (let r of ready) {
            let i = this.pendingReadRequests.indexOf(r);
            this.pendingReadRequests.splice(i, 1);
        }
        for (let r of ready) {
            r.callback();
        }
    }

    //
    // Write to Persistence
    //

    private readonly persistenceWrite = (records: RecordSet) => {
        for (let r of Object.keys(records)) {
            this.pendingWriteRequests.set(r, records[r]);
            this.doWriteIfNeeded();
        }
    }

    private readonly doWriteIfNeeded = async () => {
        if (this.pendingWriteRequests.size > 0 && !this.isWriting) {
            this.isWriting = true;
            let toWrite: RecordSet = {};
            for (let wr of this.pendingWriteRequests) {
                toWrite[wr[0]] = wr[1];
            }
            this.pendingWriteRequests.clear();

            try {
                await this.persistence.saveRecords(toWrite);
            } finally {
                this.isWriting = false;
                this.doWriteIfNeeded();
            }
        }
    }
}